### PR TITLE
Adding feature to intercept errors and save information to files

### PR DESCRIPTION
**Pull request by dchandan**
Adding feature to intercept errors encountered while (i) generating STAC items, (ii) posting STAC items to the server. Information about the dataset and the error it encountered are saved to separate log files. An example log file from a run of the stac-populator is included below.

[stac-populator_CMIP6populator_errors_20240119-185007.txt](https://github.com/crim-ca/stac-populator/files/14012790/stac-populator_CMIP6populator_errors_20240119-185007.txt)

### CODE REVIEW

**Comment by huard** [comment]
I think this PR branches off the other one about the crawler. I'll wait for you to merge the first one before reviewing.

**Comment by dchandan** [comment]
I've merged it now.

**Review by huard — COMMENTED** [review]
Maybe dumb question, but why are we using an "error file" in the first place and not just using logger to log errors to file ?

File: STACpopulator/implementations/CMIP6_UofT/add_CMIP6.py
Comment (fmigneault) [inline STACpopulator/implementations/CMIP6_UofT/add_CMIP6.py:82]: This and other above should instead make use of pythonic `raise <StacPopulatorException>("message")`. 
The parent operation then decides what to do with the exception.

File: STACpopulator/api_requests.py
Comment (fmigneault) [inline STACpopulator/api_requests.py:116]: I think it is more appropriate to raise the exception and let the parent code handle it accordingly.
For example, a 403 would make all operations useless, while 409 could be ok if override was not requested, and 502 could indicate that the endpoint is wrong entirely. 
This information is loss with this strategy.

**Comment by huard** [comment]
It's possible to configure the LOGGER object to write both to the terminal and to a file on disk. So if that meets the requirements, it could be an option. 

In another project I've used this:
```
def init_log():
    logger = logging.getLogger(__name__)
    logger.setLevel(logging.ERROR)

    fh = logging.FileHandler('log/run_MAGICC.log')
    fh.setLevel(logging.INFO)
    logger.addHandler(fh)
    return logger 
```
which logs ERRORs to the terminal, but stores all INFO messages to a file.

File: STACpopulator/populator_base.py
Comment (fmigneault) [inline STACpopulator/populator_base.py:147]: This looks a lot like a log formatter format. One should be used.

File: STACpopulator/populator_base.py
Comment (fmigneault) [inline STACpopulator/populator_base.py:145]: This is essentially what `logging.FileHandler` does, but it properly handles thread-safe operations as well.

**Review by fmigneault — CHANGES_REQUESTED** [review]
Like @huard mentioned, I think it would work better if the logger was reused. 
One way to implement it would be to add multiple log handlers. 
This way, you only need to call `log.<level>()` methods once, and each handler performs whatever it needs to do with the record respectively.

```python 
logger = get_logger("stac-populator")
log_handler = logging.StreamHandler(sys.stderr)
log_handler.setFormatter(logging._defaultFormatter)  # or whatever
err_handler = logging.FileHandler("<file-path>")
err_handler.setFormatter(logging.Formatter(fmt="{stac_item_name}: {stac_item_loc}"))
logger.addHandler(log_handler)
logger.addHandler(err_handler)

logger.debug("Custom message: {n}: {l}", n=stac_item_name, l=stac_item_loc)
```

A custom formatter could be defined to ignore the records when required arguments (stac_item_name, stac_item_loc, etc.) are missing. Just need to define the `emit` method that handles the log record.

**Comment by fmigneault** [comment]
@dchandan 
A timely video that came up in my watch list today: https://www.youtube.com/watch?v=9L77QExPmI0 
See JSON logs and extra parameter timestamps for some inspiration.

**Comment by dchandan** [comment]
Thanks for the comments and the useful video on the log file. I do like the JSON lines output format shown in the video as it would make parsing the log files for issues much easier. I've now revamped the whole app logging to use the setup in the video.

File: STACpopulator/populator_base.py
Comment (fmigneault) [inline STACpopulator/populator_base.py:155]: I suggest using `logging.INFO`/`logging.DEBUG` and use `logging.getLevelName` in the function for the configuration.

File: STACpopulator/populator_base.py
Comment (fmigneault) [inline STACpopulator/populator_base.py:149]: missing typing for `log_debug`

File: STACpopulator/populator_base.py
Comment (fmigneault) [inline STACpopulator/populator_base.py:153]: Consider `utcnow` with datetime format using `Z` or UTC timezone notation ?

File: STACpopulator/populator_base.py
Comment (fmigneault) [inline STACpopulator/populator_base.py:45]: Seems odd to me that logging would be configured here, especially since logging can be performed by other modules in the package. 
I think it should be done in the CLI where debug/log-level could be configured/modified by options: https://github.com/crim-ca/stac-populator/blob/10b07161b5690c82d88270bd39ad73e54d02340a/STACpopulator/cli.py#L159

Example of a set of options setup for the CLI: 
https://github.com/crim-ca/weaver/blob/d911c161721c2e25ee4663d34a49d2b218261710/weaver/cli.py#L1785-L1804

File: STACpopulator/logging.py
Comment (fmigneault) [inline STACpopulator/logging.py:100]: Consider `f"{JSONLogFormatter.__module__}.{JSONLogFormatter.__name__}"` so that IDEs can resolve references and make the config dynamic with the actual location.

File: STACpopulator/populator_base.py
Comment (dchandan) [reply STACpopulator/populator_base.py:156]: Sure... 🤷‍♂️ Don't see any obvious or materially significant benefit from this..

File: STACpopulator/populator_base.py
Comment (dchandan) [reply STACpopulator/populator_base.py:154]: You mean something like `datetime.strftime(datetime.utcnow(), '%Y%m%dT%H%M%S')` ?

File: STACpopulator/populator_base.py
Comment (dchandan) [reply STACpopulator/populator_base.py:46]: Doing it in the populator_base allows me to extract the name of the specific derived implementation that is being used, which I then use to name the log file.

Comment (fmigneault) [reply STACpopulator/populator_base.py:46]: The file name could be found the same way since the implementation to use is one of the required parameters of the CLI.

File: STACpopulator/populator_base.py
Comment (fmigneault) [reply STACpopulator/populator_base.py:154]: `datetime.utcnow().isoformat() + "Z"` or add the `datetime.timezone` object to the date returned by `datetime.now()`, and then `.isoformat()` that TZ-aware datetime to get something like `"2024-02-20T01:22:33.456789+00:00"`.

File: STACpopulator/populator_base.py
Comment (fmigneault) [reply STACpopulator/populator_base.py:156]: Avoid hardcoded string as much as possible to limit side-effects or having to remember (which we won't) all the places where does hardcoded string impact the code. Using predefined references makes the intention clearer.

Comment (dchandan) [reply STACpopulator/populator_base.py]: Well.... not quite. What you are saying is true, but only for strings that are hardcode with some information that is custom to the application. But, in this case, "DEBUG" or "INFO" are not going to mean anything else, like ever (or at least in our lifetimes :D). And if one ever has to change debugging levels in the app, searching for all instances of `logging.DEBUG` (and so on) is no different from searching for all instances of `"DEBUG"` (and so on). So, I think this is one of those things where the hardcoded string doesn't matter. However, it's changed in the code to use the more verbose form.

File: STACpopulator/implementations/CMIP6_UofT/add_CMIP6.py
Comment (fmigneault) [inline STACpopulator/implementations/CMIP6_UofT/add_CMIP6.py:101]: I think it should be in the parent argparser (generic one before specific implementation is called). 
`ns.debug` will not exist if using other implementations than CMIP6

File: STACpopulator/populator_base.py
Comment (fmigneault) [reply STACpopulator/populator_base.py]: Given this use case, fair enough to leave as is.

File: STACpopulator/implementations/CMIP6_UofT/add_CMIP6.py
Comment (dchandan) [reply STACpopulator/implementations/CMIP6_UofT/add_CMIP6.py]: Okay I've moved it. I'm not sure I like the current position much. Right now, one has to do:
```
stac-populator --debug run CMIP6_UofT stac-host catalog-url --auth-handler cookie --auth-identity ~/daccs_cookie.txt
```

I'd prefer if one could instead just do:

```
stac-populator run CMIP6_UofT stac-host catalog-url --auth-handler cookie --auth-identity ~/daccs_cookie.txt --debug
```

By the way, this might be related to #46. I noticed that there is something off with the CLI parser system you've put into place. I'll let you look into it since you put it in and hence you probably understand it better than me.

Comment (fmigneault) [reply STACpopulator/implementations/CMIP6_UofT/add_CMIP6.py]: To allow `stac-populator run CMIP6_UofT --debug ...`, you could define the logging options as their own `log_parser = argparse.ArgumentParser(...)` and pass that `log_parser` in `parents` (https://github.com/crim-ca/stac-populator/blob/24eeb6b0b863dd5a19f1903527300903b129f9ca/STACpopulator/cli.py#L163). Need to pass it down also to `parser_maker` (https://github.com/crim-ca/stac-populator/blob/24eeb6b0b863dd5a19f1903527300903b129f9ca/STACpopulator/cli.py#L158C32-L158C44) so it can add `log_parser` to its own `parents` as well.

**Review by fmigneault — APPROVED** [review]
@dchandan 
Good to merge, but I let you decide if you want to try out https://github.com/crim-ca/stac-populator/pull/45#discussion_r1498326083 before or not.

**Comment by dchandan** [comment]
@fmigneault 
Ok, we can do that later. Kinda pressed for time rn, so we can circle back to this later. I'll make an issue of this, so it is documented.

**Pull request merged**
